<a href="https://colab.research.google.com/github/ad-riyan/Tutorial_Python4MechanicalEngineers/blob/master/P01_Python3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python 3
*Python for Mechanical Engineers*

<br><br>

### Adriyan 

<br>
&copy; Adriyan, 2020

*Digunakan untuk keperluan pelatihan!*

<br>

---

<br>

In [0]:
!pip install version_information
%load_ext version_information
%version_information

  Created wheel for version-information: filename=version_information-1.0.3-cp36-none-any.whl size=3882 sha256=a3562d23ec80dd78e0eb76a4448ae2ac9ee8c8404fbc497c51a35cdaeaf2c641
  Stored in directory: /root/.cache/pip/wheels/1f/4c/b3/1976ac11dbd802723b564de1acaa453a72c36c95827e576321
Successfully built version-information


Software versions
Python 3.6.9 64bit [GCC 8.4.0]
IPython 5.5.0
OS Linux 4.19.104+ x86_64 with Ubuntu 18.04 bionic
Fri May 29 02:54:45 2020 UTC

In [0]:
import sys
print("Python version: %s." % (sys.version, ))

Python version: 3.6.9 (default, Apr 18 2020, 01:56:04) 
[GCC 8.4.0].


In [0]:
from IPython.display import Image, display

## 1 | Struktur bahasa pemograman Python

### 1.1 | Statement
Statement merupakan instruksi yang diberikan tiap baris untuk dijalankan oleh program.

Ketikkanlah 
```python
print("Halo, Python")
```
pada *cell* berikut ini dan kemudian tekan `Ctrl+Enter`, atau klik tombol Run cell.

### 1.2 | Variables dan Literals

Variables merupakan suatu penamaan atau pengindentifikasian yang  digunakan untuk menampung suatu data atau informasi dalam pemrograman.

Aturan penamaan variabel di python:
1.	Variabel harus dimulai dengan huruf (a-z, A-Z) atau karakter garis bawah *underscore* (_) dan tidak dapat dimulai dengan angka (0-9).
2.	Variabel hanya boleh mengandung karakter alfabet (a-z, A-Z), bilangan (0-9), dan *underscore* (_)
3.	Variabel bersifat *case-sensitive* yang menyatakan bahwa variabel TINGGI, tinggi, dan Tinggi merupakan  tiga buah variabel yang berbeda.

Literals merupakan data atau informasi yang digunakan untuk mengisi suatu variabel.

```python
bilangan1 = 10
Bilangan_2 = 3.14
Frasa = "Halo Dunia"
```


Di Python terdapat cara istimewa dalam mendeklarasikan suatu variabel yaitu:

1. Dikumpulkan semua dalam sebuah tuple.
```python
bilangan = 3, 4
print(bilangan)
```


2. Satu-satu dideklarasikan dalam satu baris
    * cara pertama
```python
bilangan1, Bilangan_2 = 3, 4
print(bilangan1)
print(Bilangan_2)
```

2. Satu-satu dideklarasikan dalam satu baris (lanjutan...)
    * cara kedua
```python
salam = "Selamat Pagi", Nama = "Python”
print(salam)
print(Nama)
```

### 1.3 | Reserved Words

*Reserved words* merupakan kata-kata yang dan dilarang digunakan untuk variables dan literals. 

In [0]:
import keyword
print("Jumlah reserved words: %d\n" % (len(keyword.kwlist),))
print(keyword.kwlist)

Jumlah reserved words: 33

['False', 'None', 'True', 'and', 'as', 'assert', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


### 1.4 | Comments

Comments merupakan kumpulan teks yang dituliskan di dalam sebuah program dan tidak akan mempengaruhi hasil dari sebuah program. 

*Comments* ialah salah satu komponen yang penting dalam pengembangan program. *Comments* dapat diselipkan dimanapun di antara *statements* yang telah dituliskan untuk:
* untuk berkomunikasi dengan rekan programmer lainnya dalam satu tim, 
* docstring dalam suatu class dan function yang ditujukan untuk help pada class dan function tersebut.

Ada dua jenis comments di dalam Python, yaitu:

a.	*single line comment* (comments dalam satu baris) seperti pada contoh berikut:

```python
# perintah pada baris ini tidak mempengaruhi program
```

b.	*multi line comment* (comments dalam beberapa baris) yang umum digunakan untuk *docstring*:
* triple single quote
```python
'''
perintah ini tidak akan dieksekusi oleh Python 
dan perintah ini juga tidak akan dieksekusi 
perintah ini juga tidak akan dieksekusi
'''
```

atau
* triple double quote
```python
"""
perintah ini tidak akan dieksekusi oleh Python 
dan perintah ini juga tidak akan dieksekusi 
perintah ini juga tidak akan dieksekusi
"""
```



## 2 | Tipe Data

Tipe data:
1. Tipe data dasar, berupa `NoneType`, `bool`, `int`, dan`float`.
2. Tipe data text sequence berupa `str`, serta tipe data sequence yaitu `list`, dan `tuple`.
3. Tipe data set, berupa `set` yang terdiri atas `set` dan `frozen set`.
4. Tipe data mapping, berupa `dict`.

### 2.1 | Tipe data dasar

Ketikkanlah perintah berikut pada cell yang disediakan
```python
print(type(None))
print(type(True))
print(type(26))
print(type(2.78))
```

#### 2.1.1 | Pengkonversian tipe data dasar
* Konversi ke tipe data `bool`:
```python
print(bool(0))
print(bool(1))
```

* Konversi ke tipe data `int`:
```python
print(int(2.4))
print(int("6"))
```

* Konversi ke tipe data `float`:
```python
print(float(2))
print(float("6.1"))
```

#### 2.1.2 | Operator

1. *Aritmetic Operators*: `+`, `-`, `*`, `/`, `//`, `%`, `**`

| Operator	| Name	        | Example   |
|:---------:|:--------------|:---------:|
| +	        |Addition	    |a + b	    |
| -	        |Subtraction	|a - b	    |
| *	        |Multiplication	|a * b	    |
| /	        |Division	    |a / b	    |
| //	    |Floor division	|a // b     |
| %	        |Modulus	    |a % b      |	
| **	    |Exponentiation	|a ** b	    |

Contoh:
```python
a, b = 8, 3
print("a + b = ", a + b)    # Penjumlahan
print("a - b = ", a - b)    # Pengurangan
print("a * b = ", a * b)    # Perkalian
print("a / b = ", a / b)    # Pembagian
print("a // b = ", a // b)  # Floor division
print("a % b = ", a % b)    # Modulo (sisa bagi)
print("a ** b = ", a ** b)  # Pemangkatan
```

2. *Comparison Operators*

|Operator|Nama	                   |Contoh     |
|:------:|:------------------------|:---------:|
|==	     |Equal	                   |a == b	
|!=	     |Not equal	               |a != b	
|>	     |Greater than             |a > b	
|<	     |Less than	               |a < b	
|>=	     |Greater than or equal to |a >= b	
|<=	     |Less than or equal to	   |a <= b

Contoh:
```python
x, y = 10, 6.2
# Buatlah keseluruhan operator tersebut setelah komentar ini
```

3. *Logical Operators*:

|Operator|Deskripsi |Contoh|
|:------:|:---------|:----:|
|and 	 |Bernilai `True` jika kedua logika bernilai `True` |a < 5 and  a < 10	
|or	     |Bernilai `True` jika salah satu logika bernilai `True`	|a < 5 or a < 2	
|not	 |Membalikkan nilai logika|not(a < 5 and a < 10)

Contoh:
```python
a = 4
print("and: ", a < 5 and a < 10)
print("or : ", a < 5 or a < 2)
print("not: ", not(a < 5 and a < 10))
```

4. *Assignment Operators*:

|Operator|Contoh|Contoh ini ekivalen dengan|
|:------:|:-----|:-------------------------|
|`=`	 |`a = 5`	|`a = 5`	
|`+=`	 |`a += 3`	|`a = a + 3`	
|`-=`	 |`a -= 3`	|`a = a - 3`	
|`*=`	 |`a *= 3`	|`a = a * 3`	
|`/=`	 |`a /= 3`	|`a = a / 3`	
|`%=`	 |`a %= 3`	|`a = a % 3`	
|`//=`	 |`a //= 3`	|`a = a // 3`	
|`**=`	 |`a **= 3` |`a = a ** 3`

5. *Membership Operators*

Untuk menguji keanggotaan suatu tipe data dasar pada suatu sequence: text, list, atau tuple.

|Operator|Descripsi|Contoh|
|:------:|:--------|:-----|
|in 	 |Bernilai `True` jika suatu sequence `b` memiliki nilai seperti yang diberikan oleh `a`|`a in b`|
|not in	 |Bernilai `True` jika suatu sequence `b` tidak memiliki nilai seperti yang diberikan oleh `a`|`a not in b`|

Contoh:
```python
a = 3
b = [2, 4, 3, 7]
print("a in b    :", a in b)
print("a not in b:", a not in b)
```

6. *Identity Operators*

Membandingkan lokasi memori dua objek 

|Operator|Descripsi|Contoh|
|:------:|:--------|:-----|
|is 	 |Bernilai `True` jika `a` dan `b` merupakan objek yang sama|`a is b`|
|is not	 |Bernilai `True` jika `a` dan `b` bukan merupakan objek yang sama|`a is not b`|

Contoh:
```python
a = 2
b = 2
print("a is b    :", a is b)
print("a is not b:", a is not b)
```

### 2.2 | Tipe data Sequence dan Text Sequence

[Tipe data sequence](https://docs.python.org/3.7/library/stdtypes.html#sequence-types-list-tuple-range)::

* Tipe data tuple `tuple`: `(e0, e1, ...)`
* Tipe data list `list`: `[e0, e1, ...]`

dengan `e0`, `e1`, ..., merupakan elemen pertama, kedua, dst.

[Tipe data text sequence](https://docs.python.org/3.7/library/stdtypes.html#text-sequence-type-str):: tipe data string `str`: `'contoh string'` atau `"contoh string"`

#### 2.2.1 | Sequence

Setiap elemen dalam tipe data sequence (tuple dan list) dapat diberikan berbagai tipe data seperti `bool`, `NoneType`, `int`, `float`, `str`, `list`, `tuple`, `set`, `dict`, `numpy.ndarray`, `pandas.DataFrame`, `pandas.Series`, atau `object dari suatu class yang telah dibangun atau berasal dari library/packages tertentu`.

1. **Tuple**, yang dapat dinyatakan secara umum dengan `my_tuple = (e0, e1, ..., ek, ..., en)`
    * Data setiap elemen seragam, misal `int` untuk semua elemen
    ```python
    tuple1 = (1, 3, 7, -6, 7)
    print(tuple1)
    # atau
    tuple1
    ```

    * Data setiap elemen tidak seragam,
    ```python
    tuple2 = (1, 'Python 3', 7.48, True, None, (1, 4, 5.6))
    tuple2
    ```

[*Built-in method for tuple*](https://docs.python.org/3.7/library/stdtypes.html#sequence-types-list-tuple-range)

2. **List**, yang dapat dinyatakan secara umum dengan `my_list = [e0, e1, ..., ek, ..., en]`
    * Data setiap elemen seragam, misal `int` untuk semua elemen
    ```python
    list1 = [1, 3, 7, -6, 7]
    list1
    ```

    * Data setiap elemen tidak seragam,
    ```python
    list2 = [1, 'Python 3', 7.48, True, None, ['Saya', 3.14], (1, 4, 5.6)]
    list2
    ```

    * *List of list*
    ```python
    list3 = [[7, 0, -4, 3],
               ['a', 9, 0, 5, -0.67],
               [4.6, True]]
    ```

[*Built-in method for list*](https://docs.python.org/3.7/library/stdtypes.html#sequence-types-list-tuple-range)

#### 2.2.2 | Text Sequence

Dapat dinyatakan dengan 

```python
string1 = "Saya belajar Python 3."
string2 = 'Belajar Python 3 mengasyikan.'
string3 = """
            Saya sedang belajar Python 3.
            Belajar Python 3 mengasyikan.
            Google colaboratory --> :)
          """
print(string1)
print(string2)
print(string3)
```

*String* dapat disambungkan dengan menggunakan operator `+` atau menggunakan method `.join()`


Operator `*` dapat digunakan untuk mereplika suatu string.

[*Built in method for str*](https://docs.python.org/3.7/library/stdtypes.html#text-sequence-type-str)

#### 2.2.3 | Mengakses elemen tipe data text sequence, dan sequence.

```python
my_str = "Python3"
my_list1 = [1, 3.14, -5, 'list', (3, 8, '$100'), True, None, -9.3]
my_list2 = [[7, 0, -4, 3, 67, -9.45, '$5'],
            ['a', 9, 0, 5, -0.67],
            [4.6, True]]
```

1. Mengakses elemen tunggal.

Untuk mengakses elemen tunggal dari tipe data text sequence, sequence dan set dapat dilakukan dengan menggunakan suatu nilai integer urutan elemen tersebut. Python urutan selalu dimulai dari `0`.

Jika ingin mengakses elemen mulai dari kanan ke kiri maka digunakan nilai integer negatif. Elemen paling kanan adalah bernilai `-1`. 

![](https://github.com/ad-riyan/Tutorial_Python4MechanicalEngineers/blob/master/figures_Python4ME/P01-SingleAccess.png?raw=1)

2. *Slicing*

Dapat dipandang sebagai proses pemotongan sesuatu seperti yang dilustrasikan berikut

![](https://github.com/ad-riyan/Tutorial_Python4MechanicalEngineers/blob/master/figures_Python4ME/P01-Slicing.png?raw=1)

Secara umum, formula untuk melakukan *slicing* pada suatu variabel `var` adalah

```python
var[start:stop:step]
```

`start` merupakan integer awal untuk slicing 

`stop` merupakan integer akhir untuk slicing

`step` merupakan integer langkah/increment untuk slicing

Bagaimana pada list of list? Kita akan cek pada `my_list2`

Untuk tuple of tuple, list of tuple, tuple of list, dan berbagai variannya adalah sama denga list of list.

**NOTE**: Kedua teknik ini juga digunakan pada `numpy` array, `pandas` DataFrame dan `pandas` Series.

### 2.4 | Tipe data Set

Merupakan tipe data yang berbeda dalam hal indeks, pengurutan, dan memiliki nilai unik.

```python
set1 = {1, 4, 6, 8, 4, 5, 6, 6}
print("Set: ", set1)
print(type(set1))
frozen_set1 = ({1, 4, 6, 8, 4, 5, 6, 6}) 
print("Frozen set: ", frozen_set1)
print(type(frozen_set1))
```

Tipe data `set` ini digunakan untuk membuat sebuah set (himpunan).

### 2.5 | Tipe data Mapping

Tipe data mapping dikenal dengan nama *dictionary* `dict` yang memiliki elemen berupa pasangan `key` dan `value`.

    my_dict = {key1: value1, key2: value2, key3: value3, ...}

`key` dan `value` dapat berupa data dengan tipe `str`, `int`, `float`, `bool`, `None`, `list`, `tuple`, `set`, `dict`, dan tipe data yang dihasilkan dari package/library seperti numpy, pandas, sympy, dll.

    my_dict = {'str': 1, True: 3, 1.34: 67, None: 45, (3, 4): 78}

Untuk keterbacaan biasanya digunakan tipe data `str` dan `int` meskipun tipe data lain dapat digunakan namum dengan kehati-hatian nantinya ketika digunakan dalam sebuah program.

Contoh:
```python
lds_dict = {'Nama': 'Laboratorium Dinamika Struktur',
            'Ka. Lab': 'Prof. Dr.-Ing. Mulyadi Bur',
            'Tahun Berdiri': 1996,
            'WA Group': 'LDS family',
            'Anggota': 111}
```


In [0]:
lds_dict = {'Nama': 'Laboratorium Dinamika Struktur',
            'Ka. Lab': 'Prof. Dr.-Ing. Mulyadi Bur',
            'Tahun Berdiri': 1996,
            'WA Group': 'LDS family',
            'Anggota': 111}
lds_dict

{'Anggota': 111,
 'Ka. Lab': 'Prof. Dr.-Ing. Mulyadi Bur',
 'Nama': 'Laboratorium Dinamika Struktur',
 'Tahun Berdiri': 1996,
 'WA Group': 'LDS family'}

Untuk mengakses elemen pada sebuah `dict` digunakan `key`-nya.

```python
my_dict[key1]
# atau
my_dict[key3]
```

Untuk mengetahui apa saja `key` yang dimiliki oleh sebuah `dict` digunakan `my_dict.keys()` dan `my_dict.values()` untuk `value` yang dimilikinya.

Menambah data ke dalam variabel bertipe data `dict`

In [0]:
lds_dict['Tambah'] = 12
lds_dict

{'Anggota': 111,
 'Ka. Lab': 'Prof. Dr.-Ing. Mulyadi Bur',
 'Nama': 'Laboratorium Dinamika Struktur',
 'Tahun Berdiri': 1996,
 'Tambah': 12,
 'WA Group': 'LDS family'}

In [0]:
lds_dict.update(Saya = 'Asisten2020')
lds_dict

{'Anggota': 111,
 'Ka. Lab': 'Prof. Dr.-Ing. Mulyadi Bur',
 'Nama': 'Laboratorium Dinamika Struktur',
 'Saya': 'Asisten2020',
 'Tahun Berdiri': 1996,
 'Tambah': 6,
 'WA Group': 'LDS family'}

## 3 | Pengkondisian (*Conditioning*)

Pemilihan keputusan berdasarkan kondisi-kondisi yang diberikan.

### 3.1 | if ... else ...

Konstruksi `if ... else ...`
```python
if condition:
    <<statement where condition fulfilled>>
else:
    <<otherwise>>
```

Jika sebuah mobile robot menerima masukan dari sensor mechanical switch yang berada di bagian depannya. Bagaimana menjadikannya mundur?

```python
mech_switch_front = True 
if mech_switch_front is True:
    print("Move backward")
else:
    print("Move forward")
```

### 3.2 | if ... elif ... else ...

Konstruksi `if ... elif ... else ...`
```python
if condition1:
    <<statement where condition1 fulfilled>>
elif condition2:
    <<statement where condition2 fulfilled>> 
elif condition3:
    <<statement where condition3 fulfilled>>
else:
    <<otherwise>>
```

Huruf mutu dari nilai seorang mahasiswa memiliki interval:

$$\begin{array}{rcl}
80 <= & A & <= 100 \\
70 <= & B & < 80 \\
60 <= & C & < 70 \\
50 <= & D & < 60 \\
40 <= & E & < 50 \\
\end{array}$$ 

### 3.3 | Nested if

Konstruksi
```python
if condition1:
    <<statement where condition1 fulfilled>>
elif condition2:
    if condition21:
        <<statement where condition21 fulfilled>> 
    elif condition22:
        <<statement where condition22 fulfilled>>
    else:
        <<otherwise>>
    <<statement where condition2 fulfilled>> 
elif condition3:
    <<statement where condition3 fulfilled>>
else:
    <<otherwise>>
```


## 4 | Perulangan (*Looping*)

Jenis perulangan dengan python:
1. For loop
2. While loop 

### 4.1 | For Loop

1. Konstruksi for loop

```python
for item in sequence:
    <<statement 1 goes here>>
    <<statement 2 goes here>>
    ...
    <<statement n goes here>>
```
`sequence` dapat berupa `tuple`, `list`, `set`, `str`, `range(n)`

Contoh:
```python
daftar_saya = [1, 4, 6.8, 'Python']
for item in daftar_saya:
    print(daftar_saya)

for item in range(5):
    print(item)

for item in range(1, 11, 3):
    print(item)
```

2. Konstruksi for loop jika mempergunakan indeks dari setiap item

```python
for index, item in enumerate(sequence):
    <<statement 1 goes here>>
    <<statement 2 goes here>>
    ...
    <<statement n goes here>>
```

Contoh: 
```python
daftar_saya = [1, 4, 6.8, 'Python']
for i, item in enumerate(daftar_saya):
    print(i, daftar_saya)
```

3. Konstruksi *nested loop*

```python
for item1 in sequence1:
    <<statement 1 goes here>>
    <<statement 2 goes here>>
    for item2 in sequence2:
        <<statement inner 1 goes here>>
        <<statement inner 2 goes here>>
        ...
        <<statement inner k goes here>>
    <<statement 3 goes here>>
    ...
    <<statement n goes here>>
```

Contoh:
```python
list_of_list = [[1, 4, 5, 7],
                ['a', 'b', 'c']]
for list_item in list_of_list:
    for item in list_item:
        print(item)
```

### 4.2 | While Loop

Konstruksi *while loop*
```python
while condition:
    <<statement 1 goes here>>
    <<statement 2 goes here>>
    ...
    <<statement n goes here>>
```

*while loop* akan dieksekusi jika `condition` bernilai `True`.

Contoh:
```python
i = 0
while i < 10:
    print(i)
    i += 1
```

## 5 | Functions

*Functions* ditujukan untuk memperingkas suatu kumpulan kode program untuk tujuan tertentu yang bersifat *reusable*.

1. Function
2. Anonymous function: lambdas


### 5.1 | Function

1. Konstruksi function dasar

```python
def nama_function(arg1, arg2, arg3, arg4):
    <<statement 1 goes here>>
    <<statement 2 goes here>>
    ...
    <<statement n goes here>>
    return out1, out2
```

yang dapat dipanggil dengan
```python
x, y = nama_function(a, b, c, d)
# atau
hasil = nama_function(a, b, c, d)
```

2. Konstruksi function dengan argumen memiliki nilai default.

```python
def nama_function(arg1, arg2, arg3=nilai_default):
    <<statement 1 goes here>>
    <<statement 2 goes here>>
    ...
    <<statement n goes here>>
    return out1, out2
```

yang dapat dipanggil dengan, 
* jika menggunakan nilai default 

```python
x, y = nama_function(a, b)
#atau
hasil = nama_function(a, b)
```

* jika menggunakan nilai default diubah 

```python
x, y = nama_function(a, b, c=nilai_lainnya)
# atau
x, y = nama_function(a, b, c)
# atau
hasil = nama_function(a, b, c=nilai_lainnya)
# atau
hasil = nama_function(a, b, c)
```

Perhatikan konstruksi fungsi berikut:

```python
def nama_function(arg1, arg2, arg3=default1, arg4=default2):
    <<statement 1 goes here>>
    <<statement 2 goes here>>
    ...
    <<statement n goes here>>
    return out1, out2
```

dapat dipanggil dengan
```python
args = (arg1, arg2)
kwargs = {"arg3": default1, "arg4": default2}

# Opsi 1:
out = nama_function(*args)

# atau Opsi 2:
out = nama_function(*args, **kwargs)
```

`args`: arguments dan `kwargs`: keyword arguments.

### 5.2 | Anonymous function: lambdas

Biasanya merupakan jenis *inline function*. 

1. Konstruksi satu argumen input: `my_cubic = lambda x: x**3`
2. Konstruksi dua argumen input: `my_power = lambda x, n: x**n`



## 6 | Library Standar Python 3 (built-in).

Silakan klik [library standar python 3](https://docs.python.org/3/library/).

```python
import math
import time
import os
import io
```

**NOTE**: Tujuan meng-*import library/packages* di python adalah agar *library/packages* dapat diakses untuk digunakan dalam *statement* yang dibuat/dikembangkan.

## 7 | File Handling



## 8 | Meng-*import Libraries/Packages* di luar *Built-in Libraries*

**NOTE**: Tujuan meng-*import library/packages* di python adalah agar *library/packages* dapat diakses untuk digunakan dalam *statement* yang dibuat/dikembangkan.

Untuk meng-*import libraries/packages* scientific/lainnya pada python dapat dilakukan dengan
* `import nama_packages`
* `import nama_packages as alias_packages`
* `import nama_packages.nama_modules as alias_modules`
* `from nama_packages import nama_modules`
* `from nama_packages import nama_modules as alias_modules`
* `from nama_packages.nama_modules import nama_child_modules as alias_child_modules`

Umumnya untuk scientific python stacks:
* Numpy: `import numpy as np`
* Matplotlib: `import matplotlib as mpl` dan `import matplotlib.pyplot as plt`
* Scipy: misal 
    * `from scipy.io import loadmat, savemat`
    * `from scipy.optimize import minimize`
    * `from scipy import linalg as scla`
    * `from scipy import signals, fft`
* Sympy: misal 
    * `import sympy as sym`
    * `from sympy.physics import mechanics as mec`

## 9 | Studi Kasus

Berdasarkan materi yang telah dipelajari gunakan python (dan *library built-in*) untuk memechakan persoalan menentukan akar suatu fungsi secara numerik dengan menggunakan 
1. metode bagi dua (*bisection*), dan
2. metode Newton-Raphson.

Diberikan sebuah fungsi 
$$f(x) = -0.6x^2 + 2.4x + 5.5 $$
untuk interval batas bawah $x_l = 3$ dan batas atas $x_u = 7$, khusus untuk metode Newton-Raphson gunakan nilai tebakan awal salah satu dari $x_l$ atau $x_u$. Proses iterasi dihentikan jika galat relatif $\varepsilon_r < 5 \%$.

**Note**: nilai numerik akar persamaan kuadrat dari fungsi 
$$f(x) = -0.6x^2 + 2.4x + 5.5 $$
adalah
$$x_1 = -1.62859017617954$$
$$x_2 =  5.62859017617954$$

Untuk menyimpan data pada setiap prose iterasi gunakan list of list yang mana list terdalam berisi 
* iterasi ke berapa
* $x_l$
* $x_u$
* $x_r$
* $f(x_r)$
* $\varepsilon_r$

untuk metode bisection, dan untuk metode Newton-Raphson
* iterasi ke berapa
* $x_0$
* $f(x_0)$
* $\varepsilon_r$

Sementara itu list terluar sejumlah iterasi yang digunakan.

## 10 | *What's Next?*

The following materials do not cover within this training.

1. Object Oriented Programming
2. Writing Efficient Python Code
3. Creating Robust Workflows
4. Software Engineering with Python
5. Using UnitTesting for Testing Python Packages



---

